xhistogram/core.py

```diff
@@ -222,54 +278,68 @@ def histogram(
-        else:
-            # reshape the array to 2D
-            # axis 0: preserved axis after histogram
-            # axis 1: calculate histogram along this axis
-            new_pos = tuple(range(-len(axis), 0))
-            c = np.moveaxis(a, axis, new_pos)
-            split_idx = c.ndim - len(axis)
-            dims_0 = c.shape[:split_idx]
-            assert dims_0 == kept_axes_shape
-            dims_1 = c.shape[split_idx:]
-            new_dim_0 = np.prod(dims_0)
-            new_dim_1 = np.prod(dims_1)
-            d = reshape(c, (new_dim_0, new_dim_1))
-        return d
+        axis = [int(i) for i in axis_normed]
 
-    all_args_reshaped = [reshape_input(a) for a in all_args_broadcast]
+    all_arrays = list(args)
+    n_inputs = len(all_arrays)
 
     if weights is not None:
-        weights_reshaped = all_args_reshaped.pop()
+        all_arrays.append(weights)
+        has_weights = True
     else:
-        weights_reshaped = None
+        has_weights = False
+
+    dtype = "i8" if not has_weights else weights.dtype
 
-    n_inputs = len(all_args_reshaped)
     bins = _ensure_bins_is_a_list_of_arrays(bins, n_inputs)
 
-    bin_counts = _histogram_2d_vectorized(
-        *all_args_reshaped,
-        bins=bins,
-        weights=weights_reshaped,
-        density=density,
-        block_size=block_size,
-    )
+    bincount_kwargs = dict(weights=has_weights, axis=axis, bins=bins, density=density)
+    if _any_dask_array(weights, *all_arrays):
+        # We should be able to just apply the bin_count function to every
+        # block and then sum over all blocks to get the total bin count.
+        # The main challenge is to figure out the chunk shape that will come
+        # out of _bincount. We might also need to add dummy dimensions to sum
+        # over in the _bincount function
+        import dask.array as dsa
+
+        # here I am assuming all the arrays have the same shape
+        # probably needs to be generalized
+        input_indexes = [tuple(range(a.ndim)) for a in all_arrays]
+        input_index = input_indexes[0]
+        assert all([ii == input_index for ii in input_indexes])
+        # keep these axes in the inputs
+        if axis is not None:
+            keep_axes = tuple([ii for ii in input_index if ii not in axis])
+        else:
+            keep_axes = ()
+        new_axes = {
+            max(input_index) + 1 + i: axis_len
+            for i, axis_len in enumerate(
+                [len(bin) - 1 for bin in bins]
+                + [
+                    1,
+                ]
+            )
+        }
+        out_index = keep_axes + tuple(new_axes)
+
+        blockwise_args = []
+        for arg in all_arrays:
+            blockwise_args.append(arg)
+            blockwise_args.append(input_index)
+
+        bin_counts = dsa.blockwise(
+            _bincount,
+            out_index,
+            *blockwise_args,
+            new_axes=new_axes,
+            meta=np.array((), dtype) ** bincount_kwargs,
```

Comment:
@rabernat this is a syntax bug, there is a missing comma between arguments. It should be 

```python
meta=np.array((), dtype), 
**bincount_kwargs,
```

This actually causes this blockwise algorithm to error for chunked inputs, because the bins never get passed in. The reason the tests still pass is because the lack of tests with explicit chunking means this code path is not really tested in most of the current tests.

I've made some hypothesis tests to cover different types of chunking but I'm currently trying to fix some other bugs they revealed.